### PR TITLE
Refactor Carousel component, reconfigure linting and update deps

### PR DESCRIPTION
**Pull request by zubhav**
### Description

**Refactor Carousel components**
- Extracted all props at top of component to avoid repeated object access via `props.xyz`
- Wrapped component functions in `useCallback` with correct deps
- Fixed `useEffect` usage by supplying necessary deps
- Avoided memory leaks by implementing missing cleanup functions for timeouts

**Reconfigured linting**
- Updated the linting setup to be inline with changes to React ecosystem (previous setup was from 4 years ago)
- Introduced eslint-plugin-react-hooks to display errors and warnings for exhaustive-deps rule

**Upgraded deps**
- Eslint, eslint-config-prettier, eslint-plugin-react

Screenshot 2022-05-03 at 11 10 45

#### Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

### Checklist: (Feel free to delete this section upon completion)

- [x] My code follows the style guidelines of this project (I have run `yarn lint`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (I have run `yarn test` and `yarn test-e2e`)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] For any breaking API changes, I have updated the type definitions in [`index.d.ts`](../index.d.ts)
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src-v5/slide.tsx [hunk 1]
```diff
@@ -168,7 +168,7 @@ const Slide = ({
         );
       }
     }
-  }, [isVisible]);
+  }, [adaptiveHeight, frameHeight, isVisible, setFrameHeight]);
```
Comment (ValGeorgiev) [inline src-v5/slide.tsx:171]: I would need to test this more, otherwise great one. Thanks for it

File: src-v5/carousel.tsx [hunk 2]
```diff
@@ -173,32 +223,63 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
     }
-  }, [currentSlide]);
+
+    return function cleanup() {
+      if (prevTimeout) {
+        clearTimeout(prevTimeout);
+      }
+      if (nextTimeout) {
+        clearTimeout(nextTimeout);
+      }
```
Comment (zubhav) [inline src-v5/carousel.tsx:269]: Avoid memory leaks

File: src-v5/carousel.tsx [hunk 3]
```diff
@@ -70,65 +113,74 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
     }
-  };
+  }, [currentSlide, moveSlide, propsSlidesToScroll, scrollMode, wrapAround]);
 
-  useEffect(() => {
-    if (typeof props.slideIndex === 'number' && !props.autoplayReverse) {
-      moveSlide(props.slideIndex);
-    }
-  }, [props.slideIndex]);
+  // Why is this needed?
```
Comment (zubhav) [inline src-v5/carousel.tsx:174]: I couldn't figure out why this useEffect was needed. @ValGeorgiev Any ideas?

File: src-v5/carousel.tsx [hunk 4]
```diff
@@ -70,65 +113,74 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
+  // useEffect(() => {
+  //   if (typeof slideIndex === 'number' && !autoplayReverse) {
+  //     moveSlide(slideIndex);
+  //   }
+  // }, [slideIndex, autoplayReverse, moveSlide]);
 
+  // Why is this needed?
```
Comment (zubhav) [inline src-v5/carousel.tsx:181]: It would also be helpful to understand this one too! @ValGeorgiev

File: src-v5/slide.tsx (hunk 1 above)
Comment (zubhav) [reply src-v5/slide.tsx]: Thanks, I updated those deps too. Didn't spot those warnings initially..

File: src-v5/slide.tsx [hunk 5]
```diff
@@ -168,7 +168,15 @@ const Slide = ({
...
-  }, [isVisible]);
+  }, [
+    adaptiveHeight,
+    customIndex,
+    frameHeight,
+    isVisible,
+    setFrameHeight,
+    slidesToShow,
+    visibleHeights
+  ]);
```
Comment (ValGeorgiev) [inline src-v5/slide.tsx:179]: @ValGeorgiev for testing

File: src-v5/carousel.tsx [hunk 6]
```diff
@@ -36,6 +73,16 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
 
+  const slidesToScroll =
+    animation === 'fade' ? slidesToShow : propsSlidesToScroll;
+  const dragThreshold = (carouselWidth.current || 0) / slidesToShow / 2;
+
+  const [slide] = getIndexes(
+    currentSlide,
+    currentSlide - slidesToScroll,
+    count
+  );
```
Comment (ValGeorgiev) [inline src-v5/carousel.tsx:84]: useCallback

File: src-v5/carousel.tsx [hunk 7]
```diff
@@ -70,65 +113,74 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
-    setTimeout(
-      () => {
-        if (!isMounted.current) return;
-        typeof to === 'number' && props.afterSlide(nextIndex);
-        !props.disableAnimation && setAnimation(false);
-      },
-      !props.disableAnimation ? props.speed || 500 : 40
-    ); // if animation is disabled decrease the speed to 40
-  };
+  const moveSlide = useCallback(
```
Comment (ValGeorgiev) [inline src-v5/carousel.tsx:116]: @ValGeorgiev to test

File: src-v5/carousel.tsx [hunk 8]
```diff
@@ -70,65 +113,74 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
+        wrapAround,
         currentSlide,
         count,
-        props.slidesToScroll
+        propsSlidesToScroll
       );
+
       moveSlide(nextPosition);
-    } else {
-      moveSlide();
```
Comment (ValGeorgiev) [inline src-v5/carousel.tsx:105]: @ValGeorgiev to test

File: src-v5/carousel.tsx [hunk 9]
```diff
@@ -70,65 +113,74 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
+        scrollMode,
+        wrapAround,
         currentSlide,
-        props.slidesToScroll
+        propsSlidesToScroll
       );
+
       moveSlide(prevPosition);
-    } else {
-      moveSlide();
```
Comment (ValGeorgiev) [inline src-v5/carousel.tsx:120]: @ValGeorgiev to test

File: src-v5/carousel.tsx (hunk 7 above)
Comment (ValGeorgiev) [reply src-v5/carousel.tsx:116]: All good

File: src-v5/carousel.tsx (hunk 8 above)
Comment (ValGeorgiev) [reply src-v5/carousel.tsx:105]: All good

File: src-v5/carousel.tsx (hunk 9 above)
Comment (ValGeorgiev) [reply src-v5/carousel.tsx:120]: All good

File: src-v5/carousel.tsx (hunk 4 above)
Comment (ValGeorgiev) [reply src-v5/carousel.tsx:181]: Added comments describing what's is the purpose of this hook

File: src-v5/slide.tsx (hunk 5 above)
Comment (ValGeorgiev) [reply src-v5/slide.tsx:179]: Thats working fine

File: src-v5/slide.tsx [hunk 10]
```diff
@@ -183,7 +191,13 @@ const Slide = ({
...
       }
     }
-  }, [adaptiveHeight, visibleHeights.current]);
+  }, [
+    adaptiveHeight,
+    frameHeight,
+    setFrameHeight,
+    slidesToShow,
+    visibleHeights
+  ]);
```
Comment (ValGeorgiev) [inline src-v5/slide.tsx:200]: This is breaking the adaptiveHeight property, I reverted it

File: src-v5/carousel.tsx [hunk 11]
```diff
@@ -171,14 +171,14 @@ export const Carousel = (props: CarouselProps): React.ReactElement => {
...
-  //   }
-  // }, [slideIndex, autoplayReverse, moveSlide]);
-
-  // Why is this needed?
+  // When user changed the slideIndex property from outside.
+  useEffect(() => {
+    if (typeof slideIndex === 'number' && !autoplayReverse) {
+      moveSlide(slideIndex);
+    }
+  }, [slideIndex, autoplayReverse]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (ValGeorgiev) [inline src-v5/carousel.tsx:179]: moveSlide is changed when some of the navigation buttons are used and cause triggering this hook. That's why we don't need it in the dependencies

File: src-v5/slide.tsx [hunk 12]
```diff
@@ -191,13 +191,7 @@ const Slide = ({
...
       }
     }
-  }, [
-    adaptiveHeight,
-    frameHeight,
-    setFrameHeight,
-    slidesToShow,
-    visibleHeights
-  ]);
+  }, [adaptiveHeight, visibleHeights.current]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (ValGeorgiev) [inline src-v5/slide.tsx:194]: breaking the adaptiveHeight property

**Comment by zubhav** [comment]
@ValGeorgiev Thanks for reviewing. Regarding the deps array for the useEffect - it's not a good idea to miss out deps and ignore the rule. If we have to do this, it probably means there are issues elsewhere in this component that we need to address. Here is an [article](https://dev.to/wkrueger/never-ignore-the-exhaustive-deps-rule-2ap8) which outlines why to avoid this.

I'll clean this PR up and merge it, but we should try to refactor afterwards to abide by the lint rule and prevent future maintenance issues we may face. Hope this makes all makes sense

**Comment by hsource** [comment]
@ValGeorgiev - thanks for the terrific library. This is the best carousel I've found after looking through 10+ libraries.

@zubhav - this is a great PR and should be merged in! It fixes a few bugs, especially with server-side-rendering where the number of slides showing doesn't update when `slideCount` changes. I extended this with a small fix in #903.

**Review by ValGeorgiev — APPROVED** [review]

**Pull request merged**